examples/ICERestart/main_RestartTest.go

```diff
@@ -0,0 +1,190 @@
+
+        flag.BoolVar(&isControlling, "controlling", false, "is ICE Agent controlling")
+        flag.Parse()
+
+        if isControlling {
+                localHTTPPort = 9000
+                remoteHTTPPort = 9001
+        } else {
+                localHTTPPort = 9001
+                remoteHTTPPort = 9000
+        }
+
+        http.HandleFunc("/remoteAuth", remoteAuth)
+        http.HandleFunc("/remoteCandidate", remoteCandidate)
+        go func() {
+                if err = http.ListenAndServe(fmt.Sprintf(":%d", localHTTPPort), nil); err != nil {
+                        panic(err)
+                }
+        }()
+
+        if isControlling {
+                fmt.Println("Local Agent is controlling")
+        } else {
+                fmt.Println("Local Agent is controlled")
+        }
+        fmt.Print("Press 'Enter' when both processes have started")
+        if _, err = bufio.NewReader(os.Stdin).ReadBytes('\n'); err != nil {
+                panic(err)
+        }
+
+        iceAgent, err = ice.NewAgent(&ice.AgentConfig{
+                NetworkTypes: []ice.NetworkType{ice.NetworkTypeUDP4},
+        })
+        if err != nil {
+                panic(err)
+        }
+
+        // When we have gathered a new ICE Candidate send it to the remote peer
+        if err = iceAgent.OnCandidate(func(c ice.Candidate) {
+                if c == nil {
+                        return
+                }
+
+                _, err = http.PostForm(fmt.Sprintf("http://localhost:%d/remoteCandidate", remoteHTTPPort), //nolint
+                        url.Values{
+                                "candidate": {c.Marshal()},
+                        })
+                if err != nil {
+                        panic(err)
+                }
+        }); err != nil {
+                panic(err)
+        }
+
+        // When ICE Connection state has change print to stdout
+        if err = iceAgent.OnConnectionStateChange(func(c ice.ConnectionState) {
+                fmt.Printf("ICE Connection State has changed: %s\n", c.String())
+        }); err != nil {
+                panic(err)
+        }
+
+        // Get the local auth details and send to remote peer
+        localUfrag, localPwd, err := iceAgent.GetLocalUserCredentials()
+        if err != nil {
+                panic(err)
+        }
+
+        _, err = http.PostForm(fmt.Sprintf("http://localhost:%d/remoteAuth", remoteHTTPPort), //nolint
+                url.Values{
+                        "ufrag": {localUfrag},
+                        "pwd":   {localPwd},
+                })
+        if err != nil {
+                panic(err)
+        }
+
+        remoteUfrag := <-remoteAuthChannel
+        remotePwd := <-remoteAuthChannel
+
+        if err = iceAgent.GatherCandidates(); err != nil { //start gathering candidates
+                panic(err) //will only find local because no STUN server
+        }
+
+        if err = iceAgent.Restart(localUfrag, localPwd); err != nil {

```

Comment:
I tested the update to Restart by just adding a few calls to it in the example/ping-pong/main.go. I'm not sure if this test is perfect, but I have confirmed that the Restart function can be called while there is a gatherer running and the peers can connect. (I just added: a fmt.Println("cancled") after calling agent.gatherCandidateCancel in agent.go in my testing to confirm it ran.)

Reply:
Actually, I just updated the agent_test.go instead, so we probably don't need this file.